File: code/services/rails-service/app/controllers/api_controller.rb
class ApiController < ApplicationController
    def initialize(service = ApiService.new)
      @service = service
      super()
      Rails.logger.info('Initialized ApiController')
    end
  
    def health
      Rails.logger.debug('Handling health endpoint')
      render json: @service.health_check, status: :ok
    end
  
    def chain
      Rails.logger.debug('Handling chain endpoint')
      render json: @service.chain_request, status: :ok
    end
  end

File: code/services/python-service/app/controllers/api_controller.py
from typing import Dict, Any, Tuple
from app.services import ApiService
import logging

logger = logging.getLogger(__name__)

class ApiController:
    """Handles HTTP request processing and response formatting."""
    def __init__(self, service: ApiService):
        self.service = service
        logger.info("Initialized ApiController")

    def health(self) -> Tuple[Dict[str, str], int]:
        """Handle health endpoint request."""
        logger.debug("Handling health endpoint")
        return self.service.health_check(), 200

    def chain(self) -> Tuple[Dict[str, Any], int]:
        """Handle chain endpoint request."""
        logger.debug("Handling chain endpoint")
        return self.service.chain_request(), 200